dali/pipeline/operator/builtin/external_source.cu

```diff
@@ -80,21 +85,18 @@ void ExternalSource<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
   } else if (state_info.copied_shared_data) {
     // make a shared pointer which will recycle buffer upon destruction. So when pipeline
     // no longer needs that buffer we can return it to the pool
-    void * ptr = tensor_list_elm.front()->raw_mutable_data();
+    void *ptr = tensor_list_elm.front()->raw_mutable_data();
     int device_id = tensor_list_elm.front()->device_id();
 
     auto tmp_capacity = tensor_list_elm.front()->capacity();
     auto tmp_shape = tensor_list_elm.front()->shape();
     auto tmp_type = tensor_list_elm.front()->type();
-    auto tmp_shr_ptr = shared_ptr<void>(ptr, [
-                          this,
-                          tensor_list = std::move(tensor_list_elm),
-                          copy_to_storage = std::move(internal_copy_to_storage)](void*) {
-                          RecycleBuffer(*(const_cast<std::list<uptr_tl_type>*>(&tensor_list)),
-                                        nullptr,
-                                        const_cast<std::list<uptr_cuda_event_type>*>
-                                                  (&copy_to_storage));
-                       });
+    RecycleFunctor funct{this, std::move(cuda_event), std::move(tensor_list_elm),
```

Comment:
Oh, this thing has no-op copy constructor with an assert. Can you try it in debug with asserts to see if it indeed doesn't happen?
Checking local build with tests should be enough.

Reply:
Double tested. Works with debug build as well.